go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -0,0 +1,633 @@
+package apbroute
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ktypes "k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	adminpolicybasedrouteapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+func (m *externalPolicyManager) processAddPolicy(routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) (*routePolicy, error) {
+
+	// it's a new policy
+	processedPolicies, err := m.processExternalRoutePolicy(routePolicy)
+	if err != nil {
+		return nil, err
+	}
+	err = m.applyProcessedPolicy(routePolicy.Name, processedPolicies)
+	if err != nil {
+		return nil, err
+	}
+	m.updateRoutePolicyInCache(routePolicy)
+	klog.Infof("Added Admin Policy Based External Route %s", routePolicy.Name)
+	return processedPolicies, nil
+}
+
+func (m *externalPolicyManager) applyProcessedPolicy(policyName string, routePolicy *routePolicy) error {
+	targetNs, err := m.listNamespacesBySelector(routePolicy.targetNamespacesSelector)
+	if err != nil {
+		return err
+	}
+	for _, ns := range targetNs {
+		cacheInfo, found := m.getNamespaceInfoFromCache(ns.Name)
+		if !found {
+			cacheInfo = m.newNamespaceInfoInCache(ns.Name)
+		}
+		err = m.applyProcessedPolicyToNamespace(ns.Name, policyName, routePolicy, cacheInfo)
+		m.unlockNamespaceInfoCache(ns.Name)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *externalPolicyManager) processDeletePolicy(routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) error {
+	targetNs, err := m.listNamespacesBySelector(&routePolicy.Spec.From.NamespaceSelector)
+	if err != nil {
+		return err
+	}
+	for _, ns := range targetNs {
+		cacheInfo, found := m.getNamespaceInfoFromCache(ns.Name)
+		if !found {
+			klog.Warningf("Attempting to delete policy %s from a namespace that does not exist %s", routePolicy.Name, ns.Name)
+			continue
+		}
+		err = m.removePolicyFromNamespace(ns.Name, routePolicy, cacheInfo)
```

Comment:
but here you are removing the policy from namespaces without locking the policy first. I would expect this kind of logic:

1. Lock policy
2. remove it from every namespace
3. now remove it from the cache

Without this, you could have another thread, lets say namespace watcher. I think this could potentially happen:
1. Thread policy enters this function starts removing policy from all of the namespaces.
2. Simultaneously a new namespace is added that is targeted by the policy. Thread namespace policy adds this policy to the namespace.
3. Thread policy now removes it from the cache.

Now you have a deleted policy being applied to a namespace. I think in the past we had a similar problem in network policy code. iirc we added a deleted bool to the policy itself, and set that first with a lock before we start deleting policies. Then the other threads can check this value to see if it should add or not.

@npinaeva may have other ideas